### PR TITLE
Staking Reward Rate Runtime API

### PR DESCRIPTION
**Pull request by rossbulat**
Addresses #13413.

- [x] Update inflation calculation.
- [ ] Abstract curve calculations & test reward rates with various `staked_percent` values.
- [ ] Update companion.

polkadot companion: https://github.com/paritytech/polkadot/pull/6764

### CODE REVIEW

File: bin/node/runtime/src/lib.rs [hunk 1]
```diff
@@ -1972,7 +1972,27 @@ impl_runtime_apis! {
...
+			NominationPools::api_points_to_balance(pool_id)
+		}
+
+		fn balance_to_points(pool_id: u32, new_funds: Balance) -> Balance {
+			NominationPools::api_balance_to_points(pool_id, new_funds)
+		}
+	}
+
+	impl pallet_staking_runtime_api::StakingApi<Block, Balance> for Runtime {
+		fn inflation_rate() -> Perquintill {
```
Comment (rossbulat) [inline bin/node/runtime/src/lib.rs:1988]: Inflation rate API added here. We use the same `ideal_stake` and `falloff` values that are plugged in to the reward curve.

File: frame/staking/Cargo.toml [hunk 2]
```diff
@@ -62,6 +63,7 @@ std = [
 	"sp-runtime/std",
 	"sp-staking/std",
 	"pallet-session/std",
+	"pallet-staking-reward-fn/std",
```
Comment (rossbulat) [inline frame/staking/Cargo.toml:66]: We are using the pallet reward curve's `calculate_inflation` method to determine the inflation. Now required in staking pallet.

File: frame/staking/src/pallet/impls.rs [hunk 3]
```diff
@@ -59,6 +59,38 @@ use super::{pallet::*, STAKING_ID};
...
+	/// this runtime API now to prepare the community to use it before rolling out PR#12970.
+	pub fn api_nominations_quota(_balance: BalanceOf<T>) -> u32 {
+		T::MaxNominations::get()
+	}
+
+	/// Returns the current inflation rate.
+	///
+	/// The staked amount as a PerThing is calculated and provided to `compute_inflation` alongside
+	/// the provided `ideal_staked` and `falloff` values.
+	pub fn api_inflation_rate(ideal_staked: Perquintill, falloff: Perquintill) -> Perquintill {
```
Comment (rossbulat) [inline frame/staking/src/pallet/impls.rs:77]: `0` is returned if active era is not determined. Else, we get the `total_issuance` and work out `staked` as a `PerThing`. Then plug the values into `compute_inflation`.

**Review by rossbulat — COMMENTED** [review]
@gpestana I've left comments where I've added the inflation API.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2412189

**Comment by rossbulat** [comment]
Also a part of #13069

File: frame/nomination-pools/runtime-api/src/lib.rs [hunk 4]
```diff
@@ -24,10 +24,19 @@ use codec::Codec;
 
 sp_api::decl_runtime_apis! {
 	/// Runtime api for accessing information about nomination pools.
+	#[api_version(1)]
```
Comment (bkchr) [inline frame/nomination-pools/runtime-api/src/lib.rs:27]: That is the default version.

File: bin/node/runtime/src/lib.rs [hunk 5]
```diff
@@ -1972,7 +1972,27 @@ impl_runtime_apis! {
...
+			Staking::api_inflation_rate(
+				Perquintill::from_rational(0_500_000_u64, 1_000_000_u64),
```
Comment (gpestana) [inline bin/node/runtime/src/lib.rs:1990]: These come across as magic numbers, should they be pulled from somewhere else or refactored as a const?

**Review by gpestana — CHANGES_REQUESTED** [review]

File: bin/node/runtime/src/lib.rs (hunk 5 above)
Comment (rossbulat) [reply bin/node/runtime/src/lib.rs:1990]: They are pulled from the `REWARD_CURVE` parameters on line 529. 

They are duplicates of those parameters. That build! macro only accepts integer literals also so I am unsure how we could store these values as variables to save repeating them.

File: frame/nomination-pools/runtime-api/src/lib.rs (hunk 4 above)
Comment (rossbulat) [reply frame/nomination-pools/runtime-api/src/lib.rs:27]: Hi Basti. This is the initial staking api, it does not exist prior to these PRs. So this is correct no?

File: bin/node/runtime/src/lib.rs (hunk 5 above)
Comment (kianenigma) [reply bin/node/runtime/src/lib.rs:1990]: Yeah might be unavoidable for now -- if we have to go through with this, make sure to document this in both places -- as in, if someone changes line 529, they should know that they have to update it here as well.

File: frame/nomination-pools/runtime-api/src/lib.rs (hunk 4 above)
Comment (kianenigma) [reply frame/nomination-pools/runtime-api/src/lib.rs:27]: What Basti means is that if you don't put `#[api_version(1)]`, it is already there as default.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2420973

**Comment by rossbulat** [comment]
bot rebase

**Comment by rossbulat** [comment]
git rebase

**Comment by rossbulat** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2423189

**Comment by rossbulat** [comment]
bot help

**Comment by rossbulat** [comment]
bot clean

**Comment by rossbulat** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2430909

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2430913

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2430899

**Comment by rossbulat** [comment]
bot fmt

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2431022

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2431105

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2431101

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2431190

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2431194

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2433216

**Comment by rossbulat** [comment]
bot rebase

**Comment by rossbulat** [comment]
bot rebase

**Comment by rossbulat** [comment]
bot rebase

**Comment by rossbulat** [comment]
bot rebase

**Comment by rossbulat** [comment]
bot rebase

**Comment by rossbulat** [comment]
bot clean

**Review by joepetrowski — CHANGES_REQUESTED** [review]
But this does not actually return the inflation. The runtime declares max inflation as a `const`:

```rust
const MAX_ANNUAL_INFLATION: Perquintill = Perquintill::from_percent(10);
```

And pays any inflation not going to stakers to the Treasury:

```rust
type RewardRemainder = Treasury;
```

`compute_inflation` actually computes the percentage of inflation that should go to stakers, i.e. some value between 0 and 1, a value of 1 indicating that 100% of `MAX_ANNUAL_INFLATION` is going to stakers. Inflation is a constant (defined as `MAX...` due to some losses in integer arithmetic), so I would expect `fn inflation_rate()` to just return this constant.

I would rename this API to say what it is actually returning, e.g. `pub fn ratio_of_inflation_to_stakers`

**Comment by rossbulat** [comment]
> compute_inflation

Thanks for the info. `compute_inflation` perhaps should be renamed to more closely represent what it is doing. The API should indeed just return the constant in this case.

**Comment by joepetrowski** [comment]
> > compute_inflation
> 
> Thanks for the info. `compute_inflation` perhaps should be renamed to more closely represent what it is doing.

Yes I agree `compute_inflation` is poorly named prior to this PR, but I think better not to dig the hole deeper.

**Comment by rossbulat** [comment]
> > > compute_inflation
> > 
> > 
> > Thanks for the info. `compute_inflation` perhaps should be renamed to more closely represent what it is doing.
> 
> Yes I agree `compute_inflation` is poorly named prior to this PR, but I think better not to dig the hole deeper.

The value I am trying to calculate is the [inflation rate](https://staking.polkadot.network/#/overview) figure on staking dashboard and JS Apps. I now suspect we can do `compute_inflation()` * MAX_INFLATION to get this figure

**Comment by joepetrowski** [comment]
You can see already on that part of the UI it is inconsistent. First it says, "Inflation Rate":

image

But if you click the info button, it tells you that inflation is fixed and a remainder goes to the Treasury. I.e. 7.27% is **not** the inflation rate:

image

`compute_inflation` does not return that. [You can see](https://github.com/paritytech/substrate/blob/e374a33/frame/staking/reward-fn/src/lib.rs#L57) that the first line just returns the ratio of actual/ideal stake if actual < ideal (where this dropoff from 1 is linear, not exponential):

```rust
pub fn compute_inflation<P: PerThing>(stake: P, ideal_stake: P, falloff: P) -> P {
	if stake < ideal_stake {
		// ideal_stake is more than 0 because it is strictly more than stake
		return stake / ideal_stake
	}
	// ...snip...
```

This just gets used in computing the [era payout](https://github.com/paritytech/polkadot/blob/6cde7fa/runtime/common/src/impls.rs#L84).

**Comment by rossbulat** [comment]
> You can see already on that part of the UI it is inconsistent. First it says, "Inflation Rate":
> 
> image
> 
> But if you click the info button, it tells you that inflation is fixed and a remainder goes to the Treasury. I.e. 7.27% is **not** the inflation rate:
> 
> image
> 
Yes, I built this UI so I am aware of the content. From what you are saying it appears that the inflation calculations provided (taken from polkadot js app's useInflation hook) are incorrect.

> `compute_inflation` does not return that. [You can see](https://github.com/paritytech/substrate/blob/e374a33/frame/staking/reward-fn/src/lib.rs#L57) that the first line just returns the ratio of actual/ideal stake if actual < ideal (where this dropoff from 1 is linear, not exponential):
> 
> ```rust
> pub fn compute_inflation<P: PerThing>(stake: P, ideal_stake: P, falloff: P) -> P {
> 	if stake < ideal_stake {
> 		// ideal_stake is more than 0 because it is strictly more than stake
> 		return stake / ideal_stake
> 	}
> 	// ...snip...
> ```
> 
> This just gets used in computing the [era payout](https://github.com/paritytech/polkadot/blob/6cde7fa/runtime/common/src/impls.rs#L84).


Yes, i already concluded from your previous comment that compute_inflation is different to what I originally suspected.

**Comment by rossbulat** [comment]
I am not sure if you are misunderstanding the purpose of the API. it is to indeed return this variable rate of inflation that does indeed depend on the parameters at play, it is not a fixed value.

@kianenigma I think your input would clarify things here.

**Comment by joepetrowski** [comment]
> I am not sure if you are misunderstanding the purpose of the API. it is to indeed return this variable rate of inflation that does indeed depend on the parameters at play, it is not a fixed value.
> 
> @kianenigma I think your input would clarify things here.

Yeah I do understand what it is returning. My issue is that that thing is not inflation.

**Comment by rossbulat** [comment]
Still want to finish

**Comment by rossbulat** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2832949

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2832934

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2833039

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2833024

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2833110

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2833125

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2833197

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2833212

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2833463

**Comment by rossbulat** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2833961

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2834439

File: frame/staking/src/pallet/impls.rs [hunk 6]
```diff
@@ -974,6 +974,46 @@ impl<T: Config> Pallet<T> {
 }
 
 impl<T: Config> Pallet<T> {
+	/// Returns the current staking reward rate.
+	pub fn api_reward_rate(
```
Comment (thiolliere) [inline frame/staking/src/pallet/impls.rs:978]: Staking is currently unaware of the reward rate, this implementation makes assumption on the implementation of `Config` trait.

The payout is abstracted through associated type `Config::EraPayout` which bounds the trait `EraPayout`
Either some assumption should be enforced on the implementation of `EraPayout` on the rate can be retrieved by calling the payout for an era of one year `payout(current_total_staked, current_total_issuance, one_year)`.
But this doesn't look so good.

Better would be to:
* either add a method to `EraPayout` trait to be able to retrieve the reward rate
* or expose the rate from the node.

Comment (rossbulat) [reply frame/staking/src/pallet/impls.rs:978]: Indeed. The calculations currently in place were based on an outdated calculation of PJS. We can simply refer to era payout to determine the inflation on a per-era basis.

**Pull request closed without merging**